fedimint-server/src/net/api.rs

```diff
@@ -296,6 +310,73 @@ impl ConsensusApi {
             .await
             .map_err(|_| ApiError::server_error("Unable send event".to_string()))
     }
+
+    pub async fn build_full_server_status(&self) -> ApiResult<FullStatus> {
+        let our_epoch = self.get_epoch_count().await;
+        // Clone to avoid holding the lock while we calculate things
+        let rejoin_at_epoch = self.rejoin_at_epoch.lock().await.clone();
+        let last_epoch_by_peer = rejoin_at_epoch.into_iter().fold(
+            HashMap::<PeerId, u64>::new(),
+            |mut acc, (epoch, peers)| {
+                let epoch = epoch + 1;
+                for peer in peers {
+                    match acc.entry(peer) {
+                        std::collections::hash_map::Entry::Occupied(mut entry) => {
+                            // Only update if the epoch is higher
+                            if epoch > *entry.get() {
+                                entry.insert(epoch);
+                            }
+                        }
+                        std::collections::hash_map::Entry::Vacant(entry) => {
+                            entry.insert(epoch);
+                        }
+                    }
+                }
+                acc
+            },
+        );
+        let peers_status = self.peer_status_client.query_all_status().await;
+        let mut peers = peers_status.keys().cloned().collect::<HashSet<_>>();
+        peers.extend(last_epoch_by_peer.keys().cloned().collect::<HashSet<_>>());
+
+        let full_peer_status = peers
+            .into_iter()
+            .map(|peer| {
+                let full_status = match (
+                    peers_status.get(&peer),
+                    last_epoch_by_peer.get(&peer).copied(),
+                ) {
+                    (Some(Err(e)), _) => {
+                        warn!(target: LOG_NET_API, %peer, "Unable to get peer status: {e:?}");
```

Comment:
Probably not a "warning" level material. Other peers being down might be not unusual, especially in larger federations.

Reply:
I think this only happens when a peer has been banned. If it's just disconnected, the we get the answer below

Reply:
That *and* when the whole server is shutting down. https://github.com/fedimint/fedimint/blob/60595f294c854a24fad47243ec4d42ad67603a59/fedimint-server/src/net/peers.rs#L218

When user presses ctrl+c or systemd sends `SIGTERM` we don't want to see panics and random errors, ideally. That's why it's important to handle such conditions gracefully.